setup.py

```diff
@@ -3,128 +3,101 @@
-become an excessively large and unnecessary dependency.
-
-Think of this library as a tiny self-contained version of Boost.Python with
-everything stripped away that isn't relevant for binding generation. Without
-comments, the core header files only require ~4K lines of code and depend on
-Python (2.7 or 3.x, or PyPy2.7 >= 5.7) and the C++ standard library. This
-compact implementation was possible thanks to some of the new C++11 language
-features (specifically: tuples, lambda functions and variadic templates). Since
-its creation, this library has grown beyond Boost.Python in many ways, leading
-to dramatically simpler binding code in many common situations.""")
+import setuptools.command.sdist
+
+DIR = os.path.abspath(os.path.dirname(__file__))
+
+# PYBIND11_GLOBAL_SDIST will build a different sdist, with the python-headers
+# files, and the sys.prefix files (CMake and headers).
+
+alt_sdist = os.environ.get("PYBIND11_GLOBAL_SDIST", False)
+
+version_py = "pybind11/_version.py"
+setup_py = "tools/setup_global.py.in" if alt_sdist else "tools/setup_main.py.in"
+extra_cmd = 'cmdclass["sdist"] = SDist\n'
+
+to_src = (
+    (version_py, "tools/_version.py.in"),
+    ("pyproject.toml", "tools/pyproject.toml"),
+    ("setup.py", setup_py),
+)
+
+
+with open(version_py) as f:
+    loc = {"__file__": version_py}
+    code = compile(f.read(), version_py, "exec")
+    exec(code, loc)
+    version = loc["__version__"]
+
+
+def get_and_replace(filename, binary=False, **opts):
+    with open(filename, "rb" if binary else "r") as f:
+        contents = f.read()
+    # Replacement has to be done on text in Python 3 (both work in Python 2)
+    if binary:
+        return string.Template(contents.decode()).substitute(opts).encode()
+    else:
+        return string.Template(contents).substitute(opts)
+
+
+# Use our input files instead when making the SDist (and anything that depends
+# on it, like a wheel)
+class SDist(setuptools.command.sdist.sdist):
+    def make_release_tree(self, base_dir, files):
+        setuptools.command.sdist.sdist.make_release_tree(self, base_dir, files)
+
+        for to, src in to_src:
+            txt = get_and_replace(src, binary=True, version=version, extra_cmd="")
+
+            dest = os.path.join(base_dir, to)
+
+            # This is normally linked, so unlink before writing!
+            os.unlink(dest)
+            with open(dest, "wb") as f:
+                f.write(txt)
+
+
+# Backport from Python 3
+@contextlib.contextmanager
+def TemporaryDirectory():  # noqa: N802
+    "Prepare a temporary directory, cleanup when done"
+    try:
+        tmpdir = tempfile.mkdtemp()
+        yield tmpdir
+    finally:
+        shutil.rmtree(tmpdir)
+
+
+# Remove the CMake install directory when done
+@contextlib.contextmanager
+def remove_output(*sources):
+    try:
+        yield
+    finally:
+        for src in sources:
+            shutil.rmtree(src)
+
+
+with remove_output("pybind11/include", "pybind11/share"):
```

Comment:
Can we add a check to the contextmanager that these aren't present before?

Reply:
We could, but they are completely temporary, and normally when you build this, it will be placed into a temp folder by pip's build-isolation, so this won't even be your local directory. (build isolation may be part of the complication of importing in setup.py)

Reply:
OK, no need, then.

Reply:
(from the code here, I thought they might be placed in the root; but of course, that doesn't really make sense)

Reply:
This _can_ happen, if you disable build isolation. But it's unlikely and you shouldn't have the files in here anyway and they get replaced during the build process, so might as well remove them anyway, IMO.